docking/dock_and_equilibrate.py

```diff
@@ -0,0 +1,387 @@
+        hgt = topology.HostGuestTopology(host_nb_bp, guest_base_top)
+        # setup the parameter handlers for the ligand
+        bonded_tuples = [
+            [hgt.parameterize_harmonic_bond, ff.hb_handle],
+            [hgt.parameterize_harmonic_angle, ff.ha_handle],
+            [hgt.parameterize_proper_torsion, ff.pt_handle],
+            [hgt.parameterize_improper_torsion, ff.it_handle]
+        ]
+        combined_bps = list(final_host_potentials)
+        # instantiate the vjps while parameterizing (forward pass)
+        for fn, handle in bonded_tuples:
+            params, potential = fn(handle.params)
+            combined_bps.append(potential.bind(params))
+        nb_params, nb_potential = hgt.parameterize_nonbonded(ff.q_handle.params, ff.lj_handle.params)
+        combined_bps.append(nb_potential.bind(nb_params))
+        guest_masses = [a.GetMass() for a in guest_mol.GetAtoms()]
+        combined_masses = np.concatenate([host_masses, guest_masses])
+
+        x0 = np.concatenate([solvated_host_coords, orig_guest_coords])
+        v0 = np.zeros_like(x0)
+        print(
+            f"SYSTEM", f"guest_name: {guest_name}", f"num_atoms: {len(x0)}",
+        )
+
+        for atom_num in constant_atoms:
+            combined_masses[atom_num - 1] += 50000
+
+        seed = 2020
+        intg = LangevinIntegrator(300.0, 1.5e-3, 1.0, combined_masses, seed).impl()
+
+        u_impls = []
+        for bp in combined_bps:
+            bp_impl = bp.bound_impl(precision=np.float32)
+            u_impls.append(bp_impl)
+
+        ctxt = custom_ops.Context(x0, v0, host_box, intg, u_impls)
+
+        # collect a du_dl calculation once every other step
+        subsample_freq = 2
+        du_dl_obs = custom_ops.FullPartialUPartialLambda(u_impls, subsample_freq)
+        ctxt.add_observable(du_dl_obs)
+
+        # insert guest
+        insertion_lambda_schedule = np.linspace(
+            max_lambda, 0.0, insertion_steps
+        )
+        calc_work = True
+        for step, lamb in enumerate(insertion_lambda_schedule):
+            ctxt.step(lamb)
+            if step % 100 == 0:
+                report_step(ctxt, step, lamb, combined_bps, host_box, u_impls, guest_name, insertion_steps, "INSERTION")
+                if not fewer_outfiles:
+                    host_coords = ctxt.get_x_t()[: len(solvated_host_coords)] * 10
+                    guest_coords = ctxt.get_x_t()[len(solvated_host_coords) :] * 10
+                    write_frame(
+                        host_coords,
+                        solvated_host_mol,
+                        guest_coords,
+                        guest_mol,
+                        outdir,
+                        guest_name,
+                        str(step).zfill(len(str(insertion_steps))),
+                        f"ins",
+                    )
+                if too_much_force(ctxt, combined_bps, host_box, u_impls, lamb):
+                    calc_work = False
+                    break
+
+        if (
```

Comment:
Note that this condition is only true for ABFE - if we do RBFE this condition would no longer hold.